### DISCUSSION TITLE
Versioning with a History Table in 2.0 and Foreign Keys

### DISCUSSION THREAD

**Discussion by PHvL** [sqlalchemy/sqlalchemy#10227] [Usage Questions] [answered]
In 1.4 I'm using the Versioned class from history_meta; migration to 2.0 seems to work fine – Thanks to zzzeek's help.

We are using alembic to generate all relations in our PostgreSQL database. If there were columns with ForeignKey constraints, these were not created on the History-tables by alembic in SQLAlchemy 1.4, but are now created on the History-tables as well using SQLALchemy 2.0.

In addition, alembic wants to use the same names for the FK constraints on the History-tables as were used n the original tables.

```python
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql.schema import MetaData
from .history_meta import Versioned

NAMING_CONVENTION = {
	"ix": "ix_%(column_0_N_label)s",
	"uq": "uq_%(table_name)s_%(column_0_N_name)s",
	"ck": "ck_%(table_name)s_%(constraint_name)s",
	"fk": "fk_%(table_name)s_%(column_0_N_name)s_%(referred_table_name)s",
	"pk": "pk_%(table_name)s",
}

Base = declarative_base(metadata=MetaData(naming_convention=NAMING_CONVENTION))

class Person(Base):
	__tablename__ = 'persons_local'
	person_id = Column(String(length=16), unique=True)
class ExamPerson(Versioned, Base):
	__tablename__ = 'exams_persons'
	exam_id = Column(Integer, ForeignKey('exams_tumonline.exam_id', ondelete='CASCADE'), primary_key=True, autoincrement=False)
	person_id = Column(String(length=16), ForeignKey(Person.person_id, onupdate="cascade"), primary_key=True)
class ExamTUMonline(Versioned, Base):
	__tablename__ = 'exams_tumonline'
	exam_id = Column(Integer, primary_key=True, autoincrement=False)
```
Running `alembic revision --autogenerate` after switching to SQLAlchemy 2.0 creates a migration with new FK constraints on the history-tables
```python
op.create_foreign_key(op.f('fk_exams_persons_person_id_persons_local'), 'exams_persons_history', 'persons_local', ['person_id'], ['person_id'], onupdate='cascade')
op.create_foreign_key(op.f('fk_exams_persons_exam_id_exams_tumonline'), 'exams_persons_history', 'exams_tumonline', ['exam_id'], ['exam_id'], ondelete='CASCADE')
```

Is it intentional that the foreign keys are copied to the history table?

This would not make sense for many of our current applications of the Versioned mixin. In the example given, rows in `exams_tumonline` and `exams_persons` linked by the foreign key are regularly deleted simultaneously (ON DELETE CASCADE), but we still want to keep the history, which the FK on exams_persons_history.exam_id would prevent.

Or did I need to do something differently in defining the models?

I'd be grateful for any help.

**Comment by zzzeek** [comment, marked as answer]
the recipe does copy FK constraints but you can modify it to not do this by having it remove any FK constraints

```diff
diff --git a/examples/versioned_history/history_meta.py b/examples/versioned_history/history_meta.py
index 806267cb41..8485528bce 100644
--- a/examples/versioned_history/history_meta.py
+++ b/examples/versioned_history/history_meta.py
@@ -90,7 +90,7 @@ def _history_mapper(local_mapper):
 
         for const in list(history_table.constraints):
             if not isinstance(
-                const, (PrimaryKeyConstraint, ForeignKeyConstraint)
+                const, PrimaryKeyConstraint
             ):
                 history_table.constraints.discard(const)
 

```

that logic is yours to edit, modify, pdb into, to make it generate whatever kind of history table you'd want.  it's likely just copying the existing constraint names, so you can change those by setting `constraint.name`

**Reply by zzzeek** [reply]
it appends constraints right here:

https://github.com/sqlalchemy/sqlalchemy/blob/701b80099931f5fb8a86a3774087399af35cc587/examples/versioned_history/history_meta.py#L126-L129

you can pdb right there and see / modify what it's doing

**Reply by PHvL** [reply]
Thanks for the super quick reply. Your first hint does the trick and seems to restore the old behavior.

This super_fks list confused me since it was also present in the 1.4 version and is not responsible for the foreign keys on the history table.
